File: bindings/js-browser/src/agent/agent.js
/**
 * agent.js – Agent: orchestrates chat, embedding, knowledge retrieval, and
 * tool execution in an agentic loop.
 *
 * The Agent:
 *  - Owns a LlamaChat engine, a LlamaEmbed engine, a KnowledgeStore, and a
 *    ToolRegistry.
 *  - Maintains per-session conversation history in memory.
 *  - On each chat() call:
 *      1. Embeds the user message and searches the knowledge store.
 *      2. Injects relevant context into the system prompt.
 *      3. Loops up to MAX_TOOL_ITERATIONS times:
 *         - If the model returns a tool call, executes it and feeds the result
 *           back as a "tool" role message.
 *         - If the model returns text, returns it to the caller.
 *
 * Usage:
 *   const agent = await Agent.create('chat.gguf', 'embed.gguf', ':memory:');
 *   await agent.addDocument('The capital of France is Paris.');
 *   agent.addTool({ name: 'lookup', ... }, async (args) => '...');
 *   const reply = await agent.chat('session-1', 'What is the capital of France?');
 *   agent.close();
 */

import { zipSync, unzipSync } from 'fflate';
import { LlamaChat }      from '../chat.js';
import { LlamaEmbed }     from '../embed.js';
import { KnowledgeStore } from '../knowledge/store.js';
import { ToolRegistry }   from '../tools/registry.js';

const MAX_TOOL_ITERATIONS = 10;

export class Agent {
  /** @type {LlamaChat}      */ #chat;
  /** @type {LlamaEmbed}     */ #embed;
  /** @type {KnowledgeStore} */ #store;
  /** @type {ToolRegistry}   */ #registry;
  /** @type {string}         */ #chatModelPath;
  /** @type {string}         */ #embedModelPath;
  /** @type {Map<string,Array>} sessionId → message history */
  #sessions = new Map();
  /** @type {boolean}        */ #closed = false;

  /**
   * @private – use {@link Agent.create} instead.
   */
  constructor(chat, embed, store, registry, chatModelPath, embedModelPath) {
    this.#chat           = chat;
    this.#embed          = embed;
    this.#store          = store;
    this.#registry       = registry;
    this.#chatModelPath  = chatModelPath;
    this.#embedModelPath = embedModelPath;
  }

  /**
   * Create an Agent that owns all of its engines and stores.
   *
   * @param {string} chatModelPath   path / URL to the chat GGUF model
   * @param {string} embedModelPath  path / URL to the embed GGUF model
   * @param {string} [storagePath]   ignored (JS store is always in-memory)
   * @param {object} [opts]
   * @param {Function} [opts.onEvent]  optional event callback forwarded to engines
   * @returns {Promise<Agent>}
   */
  static async create(chatModelPath, embedModelPath, storagePath = ':memory:', opts = {}) {
    const chat  = await LlamaChat.load(chatModelPath,  { onEvent: opts.onEvent });
    const embed = await LlamaEmbed.load(embedModelPath, { onEvent: opts.onEvent });
    const store    = new KnowledgeStore();
    const registry = new ToolRegistry();
    return new Agent(chat, embed, store, registry, chatModelPath, embedModelPath);
  }

  /**
   * Embed text and add it to the knowledge store.
   *
   * @param {string} text  document text (must not be empty)
   * @returns {Promise<void>}
   */
  async addDocument(text) {
    this._assertOpen();
    const vec = await this.#embed.embed(text);
    this.#store.add(text, vec);
  }

  /**
   * Register a callable tool.
   *
   * @param {object}   def      tool definition ({ name, description, parameters })
   * @param {Function} handler  async (args: object) => result
   */
  addTool(def, handler) {
    this._assertOpen();
    this.#registry.register(def, handler);
  }

  /**
   * Send a user message and return the assistant's final text reply.
   *
   * @param {string} sessionId  conversation identifier
   * @param {string} message    user message
   * @returns {Promise<string>}
   */
  async chat(sessionId, message) {
    this._assertOpen();

    // 1. Retrieve knowledge context.
    const context = await this._retrieveContext(message);

    // 2. Build system prompt.
    const systemPrompt = buildSystemPrompt(context);

    // 3. Append user message to session history.
    if (!this.#sessions.has(sessionId)) {
      this.#sessions.set(sessionId, []);
    }
    const history = this.#sessions.get(sessionId);
    history.push({ role: 'user', content: message });

    // 4. Agentic loop.
    const toolDefs = this.#registry.definitions();
    const msgs = buildMessages(systemPrompt, history);

    for (let i = 0; i < MAX_TOOL_ITERATIONS; i++) {
      // On the final iteration, force text mode to avoid an infinite loop in
      // case the model keeps returning tool calls (e.g. a misbehaving model or stub).
      const isLastIteration = i === MAX_TOOL_ITERATIONS - 1;
      const req = {
        messages:     msgs,
        responseMode: (toolDefs.length > 0 && !isLastIteration) ? 'tool_call' : 'text',
        tools:        toolDefs,
        toolChoice:   'auto',
      };

      const resp = await this.#chat.chat(req);

      if (resp.type === 'assistant_text') {
        history.push({ role: 'assistant', content: resp.text });
        return resp.text;
      }

      if (resp.type === 'structured_json') {
        const text = JSON.stringify(resp.json);
        history.push({ role: 'assistant', content: text });
        return text;
      }

      if (resp.type === 'tool_call') {
        if (!resp.tool_calls || resp.tool_calls.length === 0) {
          throw new Error('Agent: model returned tool_call with no tool calls');
        }

        // Add assistant message with tool call info.
        msgs.push({ role: 'assistant', content: JSON.stringify(resp.tool_calls) });

        // Execute each tool call.
        for (const tc of resp.tool_calls) {
          let resultStr;
          try {
            const result = await this.#registry.execute(tc.name, tc.arguments ?? {});
            resultStr = JSON.stringify(result);
          } catch (err) {
            resultStr = JSON.stringify({ error: err.message });
          }
          msgs.push({ role: 'tool', content: resultStr, tool_name: tc.name });
        }
        // Continue the loop with tool results in context.
        continue;
      }

      throw new Error(`Agent: unexpected response type "${resp.type}"`);
    }

    throw new Error(`Agent: exceeded ${MAX_TOOL_ITERATIONS} tool call iterations`);
  }

  /**
   * Clear the conversation history for a session.
   *
   * @param {string} sessionId
   */
  clearSession(sessionId) {
    this.#sessions.delete(sessionId);
  }

  /**
   * Release all owned resources. Safe to call multiple times.
   */
  close() {
    if (this.#closed) return;
    this.#closed = true;
    this.#chat.close();
    this.#embed.close();
    this.#store.close();
  }

  /**
   * Export a complete Agent bundle as a ZIP Blob.
   *
   * @returns {Promise<Blob>}
   */
  async export() {
    this._assertOpen();

    const docs = this.#store.all();

    const manifest = {
      version:       '1',
      created_at:    new Date().toISOString(),
      chat_model:    basename(this.#chatModelPath),
      embed_model:   basename(this.#embedModelPath),
      storage_path:  ':memory:',
      doc_count:     docs.length,
      embedding_dim: docs[0]?.embedding.length || 0,
    };

    const knowledgeJson = docs.map(d => ({ text: d.text, embedding: d.embedding }));

    const zip = zipSync({
      'manifest.json':  new TextEncoder().encode(JSON.stringify(manifest, null, 2)),
      'knowledge.json': new TextEncoder().encode(JSON.stringify(knowledgeJson)),
      'knowledge.db':   new Uint8Array(0),
    });

    return new Blob([zip], { type: 'application/zip' });
  }

  /**
   * Restore an Agent from a ZIP bundle.
   *
   * @param {Blob|File|ArrayBuffer|Uint8Array} zipData
   * @param {string} chatModelPath
   * @param {string} embedModelPath
   * @param {object} [opts={}]
   * @returns {Promise<Agent>}
   */
  static async importFrom(zipData, chatModelPath, embedModelPath, opts = {}) {
    let uint8;
    if (zipData instanceof Uint8Array) {
      uint8 = zipData;
    } else if (zipData instanceof ArrayBuffer) {
      uint8 = new Uint8Array(zipData);
    } else if (typeof Blob !== 'undefined' && zipData instanceof Blob) {
      uint8 = new Uint8Array(await zipData.arrayBuffer());
    } else {
      throw new Error('Agent.importFrom: unsupported zipData type');
    }

    const files = unzipSync(uint8);

    const manifestRaw = new TextDecoder().decode(files['manifest.json']);
    const manifest = JSON.parse(manifestRaw);
    if (manifest.version !== '1') {
      throw new Error(`Agent.importFrom: unsupported manifest version "${manifest.version}"`);
    }

    const knowledgeRaw = new TextDecoder().decode(files['knowledge.json']);
    const docs = JSON.parse(knowledgeRaw);

    const agent = await Agent.create(chatModelPath, embedModelPath, ':memory:', opts);
    agent._restoreDocs(docs);
    return agent;
  }

  /**
   * Restore documents directly into the knowledge store (used by importFrom).
   *
   * @param {Array<{text:string, embedding:number[]}>} docs
   */
  _restoreDocs(docs) {
    for (const doc of docs) {
      this.#store.add(doc.text, doc.embedding);
    }
  }

  // ─────────────────────────────────────────────────────────────────────────
  // Internal
  // ─────────────────────────────────────────────────────────────────────────

  async _retrieveContext(text) {
    try {
      const vec = await this.#embed.embed(text);
      return this.#store.search(vec, text, 5);
    } catch {
      return [];
    }
  }

  _assertOpen() {
    if (this.#closed) {
      throw new Error('Agent: agent is closed');
    }
  }
}

// ─────────────────────────────────────────────────────────────────────────────
// Helpers
// ─────────────────────────────────────────────────────────────────────────────

function buildSystemPrompt(docs) {
  if (!docs || docs.length === 0) {
    return 'You are a helpful AI assistant.';
  }
  const ctx = docs.map(d => `- ${d.text}`).join('\n');
  return `You are a helpful AI assistant.\n\nRelevant context:\n${ctx}`;
}

function buildMessages(systemPrompt, history) {
  return [
    { role: 'system', content: systemPrompt },
    ...history,
  ];
}

/** Extract the filename from a path (works for both / and \ separators). */
function basename(path) {
  if (!path) return '';
  const parts = path.replace(/\\/g, '/').split('/');
  return parts[parts.length - 1];
}

File: bindings/java/src/main/java/com/example/llama/agent/Agent.java
package com.example.llama.agent;

import com.example.llama.ChatEngine;
import com.example.llama.ChatOptions;
import com.example.llama.EmbedEngine;
import com.example.llama.LlamaException;
import com.example.llama.LoadOptions;
import com.example.llama.knowledge.KnowledgeStore;
import com.example.llama.model.*;
import com.example.llama.tools.ToolRegistry;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.*;
import java.nio.file.Files;
import java.nio.file.Path;
import java.sql.SQLException;
import java.time.Instant;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;
import java.util.zip.ZipOutputStream;

/**
 * Agentic chat loop that orchestrates:
 * <ul>
 *   <li>ChatEngine – language model inference</li>
 *   <li>EmbedEngine – text embedding for RAG</li>
 *   <li>KnowledgeStore – persistent vector + FTS knowledge base</li>
 *   <li>ToolRegistry – callable tool dispatch</li>
 * </ul>
 *
 * <p>The Agent maintains per-session conversation history and injects relevant
 * knowledge context into the system prompt on every turn.
 *
 * <p>Usage:
 * <pre>{@code
 * try (Agent agent = Agent.create("chat.gguf", "embed.gguf", ":memory:")) {
 *     agent.addDocument("The capital of France is Paris.");
 *     agent.addTool(new ToolDefinition("lookup", "...", params),
 *                   args -> Map.of("result", "..."));
 *     String reply = agent.chat("session-1", "What is the capital of France?");
 * }
 * }</pre>
 */
public final class Agent implements AutoCloseable {

    private static final int MAX_TOOL_ITERATIONS = 10;
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final ChatEngine      chat;
    private final EmbedEngine     embed;
    private final KnowledgeStore  store;
    private final ToolRegistry    registry;
    private final String          chatModelPath;
    private final String          embedModelPath;
    private final String          storagePath;

    /** sessionId → ordered message history. */
    private final Map<String, List<ChatMessage>> sessions = new ConcurrentHashMap<>();

    private volatile boolean closed = false;

    private Agent(ChatEngine chat, EmbedEngine embed,
                  KnowledgeStore store, ToolRegistry registry,
                  String chatModelPath, String embedModelPath,
                  String storagePath) {
        this.chat           = chat;
        this.embed          = embed;
        this.store          = store;
        this.registry       = registry;
        this.chatModelPath  = chatModelPath;
        this.embedModelPath = embedModelPath;
        this.storagePath    = storagePath;
    }

    /**
     * Create an Agent that owns its engines and knowledge store.
     *
     * @param chatModelPath  path to the chat GGUF model
     * @param embedModelPath path to the embedding GGUF model
     * @param storagePath    SQLite path or {@code ":memory:"}
     * @return a ready-to-use Agent
     * @throws LlamaException if an engine cannot be loaded
     * @throws SQLException   if the knowledge store cannot be opened
     */
    public static Agent create(String chatModelPath,
                               String embedModelPath,
                               String storagePath)
            throws LlamaException, SQLException {
        ChatEngine  chat  = ChatEngine.load(chatModelPath, new LoadOptions());
        EmbedEngine embed = EmbedEngine.load(embedModelPath, new LoadOptions());
        KnowledgeStore store;
        try {
            store = new KnowledgeStore(storagePath);
        } catch (SQLException e) {
            chat.close();
            embed.close();
            throw e;
        }
        ToolRegistry registry = new ToolRegistry();
        return new Agent(chat, embed, store, registry,
                         chatModelPath, embedModelPath, storagePath);
    }

    /**
     * Embed text and add it to the knowledge store.
     *
     * @param text document text (must not be empty)
     * @throws LlamaException if embedding fails
     * @throws SQLException   if the store cannot be updated
     */
    public void addDocument(String text) throws LlamaException, SQLException {
        float[] vec = embed.embed(text, new com.example.llama.EmbedOptions());
        store.add(text, vec);
    }

    /**
     * Register a callable tool.
     *
     * @param def     tool definition
     * @param handler receives decoded JSON arguments and returns a result
     */
    public void addTool(ToolDefinition def, ToolRegistry.Handler handler) {
        registry.register(def, handler);
    }

    /**
     * Send a user message and return the assistant's final text reply.
     *
     * <p>The method:
     * <ol>
     *   <li>Embeds the message and retrieves relevant knowledge context.</li>
     *   <li>Injects context into the system prompt.</li>
     *   <li>Appends the user message to the session history.</li>
     *   <li>Loops up to {@value #MAX_TOOL_ITERATIONS} times: executes tool
     *       calls and feeds results back until the model returns text.</li>
     * </ol>
     *
     * @param sessionId conversation identifier (arbitrary string)
     * @param message   user message
     * @return assistant text reply
     * @throws LlamaException if chat inference fails
     */
    public String chat(String sessionId, String message) throws LlamaException {
        if (closed) throw new LlamaException("ENGINE_CLOSED", "Agent is closed");

        // 1. Retrieve knowledge context.
        List<KnowledgeStore.Document> context = retrieveContext(message);

        // 2. Build system prompt.
        String systemPrompt = buildSystemPrompt(context);

        // 3. Update session history.
        sessions.computeIfAbsent(sessionId, k -> new ArrayList<>())
                .add(ChatMessage.user(message));

        // 4. Agentic loop.
        List<ToolDefinition> toolDefs = registry.definitions();
        List<ChatMessage> msgs = buildMessages(systemPrompt,
                sessions.get(sessionId));
        boolean justCalledTool = false;

        for (int i = 0; i < MAX_TOOL_ITERATIONS; i++) {
            // After a tool call + result, ask the model for a text answer.
            List<ToolDefinition> useTools = toolDefs;
            String responseMode = toolDefs.isEmpty() ? "text" : "tool_call";
            if (justCalledTool) {
                useTools = List.of();
                responseMode = "text";
                justCalledTool = false;
            }

            ChatRequest req = ChatRequest.builder()
                    .messages(msgs)
                    .responseMode(responseMode)
                    .tools(useTools.isEmpty() ? null : useTools)
                    .toolChoice(useTools.isEmpty() ? null : "auto")
                    .generation(new GenerationOptions(0.7f, 512, 0.95f, 40))
                    .build();

            ChatResponse resp = chat.chat(req, new ChatOptions());

            switch (resp.type == null ? "" : resp.type) {
                case "assistant_text" -> {
                    String text = resp.text != null ? resp.text : "";
                    sessions.get(sessionId).add(ChatMessage.assistant(text));
                    return text;
                }
                case "structured_json" -> {
                    String text = toJson(resp.json);
                    sessions.get(sessionId).add(ChatMessage.assistant(text));
                    return text;
                }
                case "tool_call" -> {
                    if (resp.toolCalls == null || resp.toolCalls.isEmpty()) {
                        throw new LlamaException("INFERENCE_FAILED",
                                "Agent: model returned tool_call with no tool calls");
                    }
                    msgs.add(ChatMessage.assistant(toJson(resp.toolCalls)));
                    for (ToolCall tc : resp.toolCalls) {
                        String resultJson;
                        try {
                            Object result = registry.execute(tc.name, tc.arguments);
                            resultJson = toJson(result);
                        } catch (Exception e) {
                            resultJson = "{\"error\":\"" +
                                    e.getMessage().replace("\"", "'") + "\"}";
                        }
                        msgs.add(ChatMessage.tool(tc.name, resultJson));
                    }
                    // Continue the loop — next iteration uses text mode.
                    justCalledTool = true;
                }
                default -> throw new LlamaException("INFERENCE_FAILED",
                        "Agent: unexpected response type: " + resp.type);
            }
        }

        throw new LlamaException("INFERENCE_FAILED",
                "Agent: exceeded " + MAX_TOOL_ITERATIONS + " tool call iterations");
    }

    /**
     * Clear the conversation history for the given session.
     *
     * @param sessionId session to clear
     */
    public void clearSession(String sessionId) {
        sessions.remove(sessionId);
    }

    /**
     * Export a complete Agent bundle to a ZIP archive at {@code zipPath}.
     *
     * <p>The archive contains:
     * <ul>
     *   <li>{@code manifest.json} – metadata (version, model names, doc count, etc.)</li>
     *   <li>{@code knowledge.json} – all documents with text and embeddings</li>
     *   <li>{@code knowledge.db} – raw SQLite file (empty placeholder for in-memory stores)</li>
     * </ul>
     *
     * @param zipPath path where the ZIP archive will be written
     * @throws IOException  if writing fails
     * @throws SQLException if the knowledge store cannot be queried
     */
    public void export(String zipPath) throws IOException, SQLException {
        List<KnowledgeStore.Document> docs = store.all();

        // Build manifest.
        Map<String, Object> manifest = new LinkedHashMap<>();
        manifest.put("version", "1");
        manifest.put("created_at", Instant.now().toString());
        manifest.put("chat_model", Path.of(chatModelPath).getFileName().toString());
        manifest.put("embed_model", Path.of(embedModelPath).getFileName().toString());
        manifest.put("storage_path", storagePath);
        manifest.put("doc_count", docs.size());
        manifest.put("embedding_dim",
                docs.isEmpty() ? 0
                        : (docs.get(0).embedding != null ? docs.get(0).embedding.length : 0));

        // Build knowledge array.
        List<Map<String, Object>> knowledgeList = new ArrayList<>(docs.size());
        for (KnowledgeStore.Document doc : docs) {
            Map<String, Object> entry = new LinkedHashMap<>();
            entry.put("text", doc.text);
            if (doc.embedding != null) {
                List<Float> embList = new ArrayList<>(doc.embedding.length);
                for (float f : doc.embedding) embList.add(f);
                entry.put("embedding", embList);
            } else {
                entry.put("embedding", List.of());
            }
            knowledgeList.add(entry);
        }

        byte[] manifestBytes  = MAPPER.writerWithDefaultPrettyPrinter()
                .writeValueAsBytes(manifest);
        byte[] knowledgeBytes = MAPPER.writerWithDefaultPrettyPrinter()
                .writeValueAsBytes(knowledgeList);

        // Read SQLite file bytes (or empty placeholder for :memory:).
        byte[] dbBytes;
        if (":memory:".equals(storagePath)) {
            dbBytes = new byte[0];
        } else {
            Path dbPath = Path.of(storagePath);
            dbBytes = Files.exists(dbPath) ? Files.readAllBytes(dbPath) : new byte[0];
        }

        // Write ZIP archive.
        try (ZipOutputStream zos = new ZipOutputStream(
                new BufferedOutputStream(Files.newOutputStream(Path.of(zipPath))))) {
            writeZipEntry(zos, "manifest.json", manifestBytes);
            writeZipEntry(zos, "knowledge.json", knowledgeBytes);
            writeZipEntry(zos, "knowledge.db", dbBytes);
        }
    }

    /**
     * Import an Agent from a previously exported ZIP archive.
     *
     * <p>The method extracts the SQLite database from the archive, creates
     * chat and embed engines from the supplied model paths, and returns a
     * fully initialised Agent.
     *
     * @param zipPath        path to the ZIP archive
     * @param chatModelPath  path to the chat GGUF model
     * @param embedModelPath path to the embedding GGUF model
     * @param storagePath    SQLite path for the restored database;
     *                       defaults to {@code "agent.db"} if {@code null} or empty
     * @return a restored Agent
     * @throws IOException    if reading the archive fails
     * @throws LlamaException if an engine cannot be loaded
     * @throws SQLException   if the knowledge store cannot be opened
     */
    @SuppressWarnings("unchecked")
    public static Agent importFrom(String zipPath, String chatModelPath,
                                   String embedModelPath, String storagePath)
            throws IOException, LlamaException, SQLException {
        if (storagePath == null || storagePath.isEmpty()) {
            storagePath = "agent.db";
        }

        // Read ZIP entries.
        byte[] manifestBytes  = null;
        byte[] dbBytes        = null;

        try (ZipInputStream zis = new ZipInputStream(
                new BufferedInputStream(Files.newInputStream(Path.of(zipPath))))) {
            ZipEntry entry;
            while ((entry = zis.getNextEntry()) != null) {
                switch (entry.getName()) {
                    case "manifest.json"  -> manifestBytes = zis.readAllBytes();
                    case "knowledge.db"   -> dbBytes = zis.readAllBytes();
                    default -> { /* skip unknown entries */ }
                }
                zis.closeEntry();
            }
        }

        // Validate manifest.
        if (manifestBytes == null) {
            throw new IOException("Invalid agent archive: missing manifest.json");
        }
        Map<String, Object> manifest = MAPPER.readValue(manifestBytes, Map.class);
        String version = String.valueOf(manifest.get("version"));
        if (!"1".equals(version)) {
            throw new IOException("Unsupported agent archive version: " + version);
        }

        // Restore SQLite database.
        if (dbBytes != null && dbBytes.length > 0) {
            Files.write(Path.of(storagePath), dbBytes);
        }

        // Create engines and knowledge store.
        ChatEngine chat = ChatEngine.load(chatModelPath, new LoadOptions());
        EmbedEngine embed = EmbedEngine.load(embedModelPath, new LoadOptions());
        KnowledgeStore store;
        try {
            store = new KnowledgeStore(storagePath);
        } catch (SQLException e) {
            chat.close();
            embed.close();
            throw e;
        }
        ToolRegistry registry = new ToolRegistry();
        return new Agent(chat, embed, store, registry,
                         chatModelPath, embedModelPath, storagePath);
    }

    /**
     * Close all owned resources (engines + knowledge store).
     * Safe to call multiple times.
     */
    @Override
    public void close() {
        if (closed) return;
        closed = true;
        store.close();
        embed.close();
        chat.close();
    }

    // ──────────────────────────────────────────────────────────────────────
    // Helpers
    // ──────────────────────────────────────────────────────────────────────

    private List<KnowledgeStore.Document> retrieveContext(String text) {
        try {
            float[] vec = embed.embed(text, new com.example.llama.EmbedOptions());
            return store.search(vec, text, 5);
        } catch (Exception ignored) {
            return List.of();
        }
    }

    private static String buildSystemPrompt(List<KnowledgeStore.Document> docs) {
        if (docs.isEmpty()) {
            return "You are a helpful AI assistant.";
        }
        StringBuilder sb = new StringBuilder("You are a helpful AI assistant.\n\nRelevant context:\n");
        for (KnowledgeStore.Document d : docs) {
            sb.append("- ").append(d.text).append('\n');
        }
        return sb.toString();
    }

    private static List<ChatMessage> buildMessages(String systemPrompt,
                                                   List<ChatMessage> history) {
        List<ChatMessage> msgs = new ArrayList<>(history.size() + 1);
        msgs.add(ChatMessage.system(systemPrompt));
        msgs.addAll(history);
        return msgs;
    }

    private static String toJson(Object obj) {
        try {
            return MAPPER.writeValueAsString(obj);
        } catch (Exception e) {
            return "{}";
        }
    }

    private static void writeZipEntry(ZipOutputStream zos, String name, byte[] data)
            throws IOException {
        zos.putNextEntry(new ZipEntry(name));
        zos.write(data);
        zos.closeEntry();
    }
}
